lib/ExpressionUtils/core/exprMatrix_utils.py

```diff
@@ -203,3 +207,117 @@ def get_expression_matrix(self, params):
+
+        dem_dict = {}
+        for i in range( 0, n_rows ):
+            dem_dict[ row_ids[i] ] = [ vals[i][0], vals[i][2] ] # [fc,q]. (not bothering to check for dups here)
+
+        return dem_dict
+
+    # implments get_enhancedFilteredExpressionMatrix() method
+
+    def get_enhancedFEM( self, params ):
+
+        if not params.get( 'fem_object_ref' ):
+            raise ValueError( "fem_object_ref parameter not given to get_enhancedFilteredExpressionMatrix" )
+
+        fem_object_ref = params.get( 'fem_object_ref' )
+
+        fem_obj_ret = self.ws_client.get_objects2(
+                       {'objects': [{'ref': fem_object_ref }]})['data'][0]
+        fem = fem_obj_ret.get( 'data' )
+        prov = fem_obj_ret.get( 'provenance')[0]
+
+        # (1) create the enhanced FEM, starting with the FEM
+
+        efem = {}
+        for k in [ 'genome_ref', 'scale', 'type' ]:
+            efem[k] = fem.get( k )
+
+        efem['data'] = {}
+        efem['data']['col_ids'] = [ "description", 
+                                    "fold-change",
+                                    "q-value",
+                                    "min",
+                                    "max",
+                                    "mean",
+                                    "std_dev",
+                                    "is_missing_values" ]
+        efem['data']['column_labels'] =[ "Description", 
+                                         "Fold change",
+                                         "Q value",
+                                         "Min. expression",
+                                         "Max. expression",
+                                         "Mean expression",
+                                         "Std. dev.",
+                                         "Missing values?" ]
+        fm = fem.get('data')
+        efem['data']['row_ids'] = fm.get('row_ids')
+        efem['data']['values' ] = []
+        n_efem_rows = len( efem['data']['row_ids'] )
+        fvals = fm.get('values')
+        if ( len( fvals ) != n_efem_rows ):
+            raise Exception( "length discrepancy in filtered expression matrix: {0} row_ids but {1} values".format( n_efem_rows, len( fvals ) ) )
+
+        for i in range( 0, n_efem_rows ):
+            efem['data']['values'].append( [ 'NA', 'NA', 'NA' ] + self.get_matrix_stats( fvals[i] ) )
+
+        # (2) Get genome object and feature descriptions, put those in column 1
+
+        feat_dict = self.gaa.get_feature_functions( { 'ref': fem.get( 'genome_ref' ), 'feature_id_list': None } )
+        for i in range( 0, n_efem_rows ):
+            desc = feat_dict.get( efem['data']['row_ids'][i] )
+            if desc:
+                efem['data']['values'][i][0] = desc     # leave as 'NA' if no entry in feat_dict
+
+        # (3) get DEM from provenance and merge the FC and q-values
+
+        if prov.get( 'input_ws_objects' ):
```

Comment:
this is a major issue. use resolved_ws_objects, not input: https://ci.kbase.us/services/ws/docs/saveobjects.html#save-an-object-with-provenance

Reply:
I'm kind of confused why we're fetching objects via provenance anyway. If you need object B to be able to use object A, a ref to object B should be embedded in object A.

Reply:
Will change to the resolved.

The whole story of diving into provenance is to get the specific Differential Expression Matrix that was used to filter the Expression Matrix  and create a filtered expression matrix.   The filtered expression matrix provenance pointed only to the Differential Expression Matrix set, which isn't enough information for this app to do its job.

Reply:
So yeah, sounds like the types need a rethink.

Reply:
Gavin, can you check to ensure I've made the change to "resolved_ws_obj"?  Jim is going to need this additional function for one of his RNAseq tickets.

Reply:
A search for `input_ws` finds nothing and this line has been switched to `resolved_ws_objects`, so you've made the change.

A couple notes:

1) This function assumes that a) there's only one provenance action in the provenance record (which is usually but not necessarily true) and b) the first item in the list of workspace objects in the provenance is the object you want. For the latter case it's not clear at all that that's always true, which is why I was recommending a reference in the object which can provide semantics for the reference.

2) In the standup we decided to discuss this further in slack.